File: 0039_combinationSum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        output = set()
        path = []
        summ = 0
        visited = set()
        candidates = list(set(sorted(candidates)))
        self.dfs(sorted(candidates, reverse = True), target, output, path, summ)
        return output
    
    def dfs(self, candidates, target, output, path, summ):
        if summ == target:
            output.add(tuple(sorted(path)))
            return
        if summ > target:
            return
        for i in range(len(candidates)):
            div = target // candidates[i]
            for j in range(1, div+1):
                self.dfs(candidates[i+1:], target, output, path + [candidates[i]] * j, summ + candidates[i] * j)

File: 0001-0500/0039_combinationSum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> output = new ArrayList<>();
        backtrack(candidates, target, output, new ArrayList<>(), 0, 0);
        return output;
    }

    public void backtrack(int[] candidates, int target, List<List<Integer>> output, List<Integer> path, int sum, int idx) {
        if (sum == target) {
            output.add(new ArrayList<>(path));
        }

        for (int i = idx; i < candidates.length; i++) {
            if (sum + candidates[i] <= target) {
                path.add(candidates[i]);
                backtrack(candidates, target, output, path, sum + candidates[i], i);
                path.remove(path.size() - 1);
            }
        }
    }
}
